JSON schema

magoni opened the pull request:
This is the initial implementation of a JSON schema to describe GOFS-lite, in order to help producers do basic sanity tests on their own feeds as requested by @richardmtl.

See README within this PR for some disclaimers (this schema is not yet complete nor sufficiently strict to fully ensure correctness, but it can help with finding basic structural errors)

Future improvements: make this more complete, strict, and implement a single source of truth for the GOFS-lite reference and this schema so that they don't drift in the future.

magoni commented:
@E-Baird following our discussion i added back `version` per the spec

E-Baird commented on schema/operating_rules.json:
```diff
@@ -0,0 +1,113 @@
+                "description": "Array of vehicle types used for delivering the on-demand service.",
+                "type": "array"
+              },
+              "fare_id": {
+                "type": "string"
+              }
```
I'm testing with some real data and the results are failing because it's got the optional fare ID field as`"fare_id": null`. The kind of small error we'd likely never catch unless/until it broke something.
I point it out not so much because I think you need to change anything in this PR, but rather just to highlight how useful of a tool this schema will be.

E-Baird commented on schema/operating_rules.json:
```diff
@@ -0,0 +1,113 @@
+                  "labor_day"
+                ],
+                "brand_id": "large_ride",
+                "vehicle_type_id": "large_van",
```
In the spec, this field is listed as an Array, but our own example shows it as a string. Can we update the example so that it's up to spec?

magoni replied:
Good catch, resolved in 1aa6415b07655d75f5fe092a401f3a316ece34b2

magoni replied:
Just updated in `reference.md` as well

E-Baird commented on schema/vehicle_types.json:
```diff
@@ -0,0 +1,55 @@
+  "$id": "https://github.com/GOFS-lite/GOFS-lite/blob/main/reference.md#vehicle_typesjson",
+  "description": "This file contains operating rules enabling on-demand services between zones or within the same zone, according to time windows and calendars. At least one operating rule MUST be defined. If start_pickup_window, end_pickup_window, and end_dropoff_window are not provided, it is assumed that the on-demand service is available at any hours of the day.",
+  "type": "object",
+  "properties": {
```
This one is still missing a `version` field

magoni replied:
Thanks! Resolved in [f3a7f50](https://github.com/GOFS-lite/GOFS-lite/pull/18/commits/f3a7f509172b240ad7c91a1fdcc701244d2db0f4)

magoni commented:
Ugh I ran into an example feed where in `booking_rules`, "from_zone_id" was defined instead of `from_zone_ids` (plural), but it still passes because the schema allows unknown fields, and it doesn't yet cover _conditionally required_s. We should probably prohibit unknown fields wherever possible to catch errors like this

magoni commented:
I've added `"additionalProperties": false` everywhere and fixed some missing properties in `booking_rules`, it's already a bit more helpful when validating!

E-Baird reviewed (approved):
Tested with a few more examples from real data, and it looks good. Stuff that should pass is passing, and it's catching errors that would otherwise be hard to find manually.

Really excited to have this in the toolkit 🎉
